### PR TITLE
MAGENTO-17: Implement webapi route for Notification

### PR DESCRIPTION
**Pull request by grandel-withreach**
- Created a new webapi route for handling the Reach Notifications
- Needed override the `DeserializerFactory` in order to be able to handle the `x-www-form-url-encoded` request that is coming from the Checkout API
- Replaced the `Notify` URL with the webapi route endpoint
- Removed the `Callback` Controller

### CODE REVIEW

File: etc/di.xml [hunk 1]
```diff
@@ -12,8 +12,38 @@
...
 
+    <preference for="Reach\Payment\Api\NotificationInterface" type="Reach\Payment\Model\ReachNotification" />
+
     <preference for="Reach\Payment\Api\Data\StashResponseInterface" type="Reach\Payment\Model\Api\Data\StashResponse" />
     <preference for="Reach\Payment\Api\Data\DutyResponseInterface" type="Reach\Payment\Model\Api\Data\DutyResponse" />
     <preference for="Reach\Payment\Api\Data\ResponseInterface" type="Reach\Payment\Model\Api\Data\Response" />
     <preference for="Reach\Payment\Api\Data\HttpResponseInterface" type="Reach\Payment\Model\Api\Data\HttpResponse" />
+    <preference for="Reach\Payment\Api\Data\NotificationResponseInterface" type="Reach\Payment\Model\Api\Data\NotificationResponse" />
+
+    <type name="Magento\Framework\Webapi\Rest\Request\DeserializerFactory">
```
Comment (grandel-withreach) [inline etc/di.xml:23]: This is overriding the default Magento `DeserializerFactory`. Needed to add an `item` for all the other types as well (`application/json`, `application/xml`, etc) using the default Magento classes as well, otherwise it wouldn't know how to handle them.

File: Webapi/Rest/Request/Deserializer/XWwwFormUrlencoded.php [hunk 2]
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace Reach\Payment\Webapi\Rest\Request\Deserializer;
+
+
+use Magento\Framework\App\State;
+use Magento\Framework\Phrase;
+use Reach\Payment\Model\PostbackNotification\Decoder;
+
+class XWwwFormUrlencoded implements \Magento\Framework\Webapi\Rest\Request\DeserializerInterface
```
Comment (grandel-withreach) [inline Webapi/Rest/Request/Deserializer/XWwwFormUrlencoded.php:10]: A lot of this content follows the default `Deserializer` classes native to Magento. Reference: https://github.com/pepe1518/magento2/blob/master/vendor/magento/framework/Webapi/Rest/Request/Deserializer/Json.php

File: Model/Cc.php [hunk 3]
```diff
@@ -601,10 +601,7 @@ protected function getConsumerInfo($order)
      */
     private function getCallbackUrl($order)
```
Comment (lizzyWithReach) [inline Model/Cc.php]: would it be a hassle to also rename this getter? 

or is this an expected function name that we shouldn't change? 
(ex expected by Magento to look like this, in which case should we adjust our terminology from notification to callback in `rest/default/V1/reach/notification`?)

File: Model/ReachNotification.php [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+     * @param $signature
+     * @return ResponseInterface
+     */
+    public function handleNotification($request, $signature)
+    {
+        try {
+            if ($this->validate($request, $signature)) {
+                // TODO:: Handle other Notification order states
+                $response = json_decode($request, true);
+                if (isset($response['OrderState']) && $response['OrderState'] == "Processed" && $response['UnderReview'] === false && !count($response['Refunds'])) {
```
Comment (lizzyWithReach) [inline Model/ReachNotification.php]: could we make this line more readable? it's a little long and also we could wait to do this until we get a linter/checkstyle in place

File: Model/ReachNotification.php [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        $this->_orderFactory = $orderFactory;
+        $this->orderRepository = $orderRepository;
+    }
+
+    /**
+     * @param $request
+     * @param $signature
+     * @return ResponseInterface
+     */
+    public function handleNotification($request, $signature)
```
Comment (lizzyWithReach) [inline Model/ReachNotification.php:53]: is it magento magic that takes the notification, uses webapi.xml to find this method then split the body into the request and signature parts for what this method is expecting as parameters?

**Review by lizzyWithReach — APPROVED** [review]
one question and two nitpicks, thanks for the comments Grandel! helped a ton!

File: Model/Cc.php (hunk 3 above)
Comment (grandel-withreach) [reply Model/Cc.php]: This method is only ever called in this class so we could change it to whatever we like

I also tested it locally and confirmed that changing it doesn't have any negative impacts. 

What do you think of `getNotifyUrl`? Or would `getNotificationUrl` be more fitting?

File: Model/Cc.php [hunk 6]
```diff
@@ -599,12 +599,9 @@ protected function getConsumerInfo($order)
      * @param object $order
      * @return string
      */
-    private function getCallbackUrl($order)
+    private function getNotifyUrl($order)
```
Comment (lizzyWithReach) [inline Model/Cc.php:602]: oh this is great! thanks! I was good with either Notify or Notification especially since it's only used in this class

File: Model/ReachNotification.php [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+        try {
+            if ($this->validate($request, $signature)) {
+                // TODO:: Handle other Notification order states
+                $response = json_decode($request, true);
+
+                $isProcessed = isset($response['OrderState']) && $response['OrderState'] == "Processed";
+                $isNotUnderReview = $response['UnderReview'] === false;
+                $hasNoRefunds = !count($response['Refunds']);
+
+                if ($isProcessed && $isNotUnderReview && $hasNoRefunds) {
```
Comment (lizzyWithReach) [inline Model/ReachNotification.php:64]: oh I find this much clearer, thank you!!

**Review by lizzyWithReach — APPROVED** [review]

File: Model/ReachNotification.php (hunk 5 above)
Comment (grandel-withreach) [reply Model/ReachNotification.php:53]: Yeah pretty much, the configuration defined in the `webapi.xml` for this route what should be ran,

```
<route url="/V1/reach/notification" method="POST">
        <service class="Reach\Payment\Api\NotificationInterface" method="handleNotification"/>
        <resources>
            <resource ref="anonymous" />
        </resources>
    </route>
```

So looking at this route configuration, we've defined that for this endpoint I want it to use the `NotificationInterface` as the Service and call the `handleNotification` method to handle the request. 

This `ReachNotification` class implements the `NotificationInterface` which defines that this class should have a `handleNotification` class that receives the two parameters. To my understanding, its Magento magic take looks at the incoming requests and checks to see if the request body contains the `request` and `signature` makes it available through those params. 

The request must also match the parameters defined in the `NotificationInterface`. DId a quick test and changes `$request` to `$test` and it failed, because my request body did not contain `test`

Comment (lizzyWithReach) [reply Model/ReachNotification.php:53]: love it! thank you for confirming! Especially when there's a bit of magic it helps to know my assumptions line up with what appears to be happening.
